demo_nodes_cpp/test/test_executables_tutorial.py.in

```diff
@@ -3,62 +3,70 @@
-        name = 'test_executable_' + str(i)
-        # The last executable is taken to be the test program (the one whose
-        # output check can make the decision to shut everything down)
-        is_last_executable = (i == (len(executables) - 1))
+    return launch_description, locals()
 
-        action = launch_test.add_fixture_action(
-            launch_description, ExecuteProcess(
-                cmd=[exe],
-                name=name,
-                output='screen'
-            ), exit_allowed=[0] if is_last_executable else True
-        )
-        launch_test.add_output_test(
-            launch_description, action,
-            output_test=create_output_test_from_file(output_file),
-            output_filter=output_filter,
-            side_effect='shutdown' if is_last_executable else None
-        )
 
-    launch_service = LaunchService()
-    launch_service.include_launch_description(launch_description)
-    rc = launch_test.run(launch_service)
+class TestExecutablesTutorial(unittest.TestCase):
+
+    def test_processes_output(self, proc_output, processes_under_test):
+        """Test all processes output against expectations."""
+        from launch_testing.tools.output import get_default_filtered_prefixes
+        output_filter = launch_testing_ros.tools.basic_output_filter(
+            filtered_prefixes=get_default_filtered_prefixes() + [
+                'service not available, waiting again...'
+            ],
+            filtered_rmw_implementation='@rmw_implementation@'
+        )
+        output_files = '@DEMO_NODES_CPP_EXPECTED_OUTPUT@'.split(';')
+        for process, output_file in zip(processes_under_test, output_files):
+            proc_output.assertWaitFor(
+                expected_output=launch_testing.tools.expected_output_from_file(
+                    path=output_file
+                ), process=process, output_filter=output_filter, timeout=10
+            )
 
-    assert rc == 0, \
-        "The launch file failed with exit code '" + str(rc) + "'. " \
-        'Maybe the client did not receive any messages?'
 
+@launch_testing.post_shutdown_test()
+class TestExecutablesTutorialAfterShutdown(unittest.TestCase):
 
-if __name__ == '__main__':
-    test_executable()
+    def test_last_process_exit_code(self, proc_info, processes_under_test):
+        """Test last process exit code."""
+        # TODO(hidmic): fail the test on a nonzero exit code.
+        #               Currently pre shutdown tests are terminating too soon and
+        #               the signal seems to be sent in the middle of the fixture
+        #               nodes' shutdown (for those that shutdown themselves) and
+        #               these are not handling this situation gracefully. If pre
+        #               shutdown test termination is delayed instead, the testing
+        #               framework complains about the launch run terminating before
+        #               tests are done.
```

Comment:
Is this an issue with `launch_testing` or with the executables?
I guess in any case it would be good to open an issue and add a link to the issue here.

Reply:
The test framework expects to be the thing that terminates the launch.  I did it this way because of my own perception of a `rostest` (from ros1) limitation where it wouldn't necessarily notice if the process under test died out from under the test.  I'm operating under the assumption that most of the time processes under test don't spontaneously die, and if they do it's a test failure.

In the case where the processes under test are expected to terminate themselves, you can add an extra launch_testing.utils.KeepAliveProc to the launch description.  This is a signal to the test framework and future readers that all of the other processes are expected to terminate on their own.

For this particular example, I would add a KeepAliveProc to the launch description, and a pre-shutdown test that awaits the shutdown of all other processes under test.  There's an [assertWaitForShutdown](https://github.com/ros2/launch/blob/0336e1677e2c7777ff3222816a8dbaaa58abf6cc/launch_testing/launch_testing/proc_info_handler.py#L103) method on the `ProcInfo` object.

Reply:
Thanks for the tip!